File: CompetitiveProgramming/DMOJ/Uncategorized/Flip the Switches/2021-07-13/FlipTheSwitches.java
/*
//////////////////////////////////////////////////////////////////////////////////
//////////////////////////// NEW REGEX SYNTAX SUMMARY ////////////////////////////
1. Match ending 0s.
    "0+$": ending with one or more 0s.
        "0+": one or more 0s.
        "$": ending with.

///////////////////////////////////////////////////////////////////////////
//////////////////////////// Flip the Switches ////////////////////////////
A security system is composed of "n" switches, where "0" represents off and "1" represents on.
Goal: deactivate the security system by turning off all switches.
You are only allowed to flip the first "k" switches at once. You may choose your own "k".
Find the minimum number of times switches need to be flipped to turn off the security system.

Sample Input
4
1001

Sample Input Explanation
Line 1: n, number of switches, i.e., number of characters in the following string.
Line 2: string of switch states.

Sample Output
3

////////////////////////////////////////////////////////////////////////
//////////////////////////// SOLUTION LOGIC ////////////////////////////
Use regular expression to count the number of matches to groups of 0s and 1s.
Trailling 0 must be removed from input string since they do not need to be flipped and will cause an erraneous answer when matched by the regular expression.
Java does not have a built-in trim end method like Python's .rstrip().
    An alternative is to match the last group of 0s with regular expression and replace ith with an empty string.
*/

import java.util.Scanner;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

/**
 * Program determines the minimum number of flip switches to turn off a security system.
 * @author Vincent Zhang
 * @since 13 July 2021
 */
public class FlipTheSwitches
{
    /**
     * Main method to solve the problem.
     * @param args Command-line arguments
     * @return void
     */
    public static void main(String[] args)
    {
        // Instantiate Scanner. Number of switches is unimportant for this approach.
        Scanner sc = new Scanner(System.in); sc.nextLine();

        // Remove trailling 0s by matching with regex.
        String s = sc.nextLine().replaceAll("0+$", "");
        
        // Close scanner to prevent wasted resources.
        sc.close();

        /*
        // Java 9+. ".results()" is the Java 9+ component where it turns the matches into a Stream.
        System.out.println
        (
            Pattern.compile("0+|1+").matcher(s).results().count()
        );
        */

        // Java 8+.
        Matcher matcher = Pattern.compile("0+|1+").matcher(s);
        int count = 0; while(matcher.find()) count++;
        System.out.println(count);
    }
}

File: CompetitiveProgramming/DMOJ/Uncategorized/Flip the Switches/2021-07-13/FlipTheSwitches.py
"""
########################################################################
############################ SYNTAX SUMMARY ############################
1. Find All Matches
    re.findall(regex, string)
    returns list

2. Regex Syntax
    "0+|1+":
        "0+": one or more 0s.
        "1+": one or more 1s.
        "|": logical or.

3. Remove Trailling Characters
    print: "1,,0.,a".rstrip(".,0a")
    output: "1"

###########################################################################
############################ Flip the Switches ############################
"n" distinct switches make up a security system. Switch state denoted by 0 (off) and 1 (on).
Goal: turn off security system by turnning off all switches.
Tool: can only flip the first k switches together, where "k" is decided by the program.
Output the minimum number of times the switches need to be flipped to turn off the security system.

Sample Input
4
1001

Sample Input Explanation
Line 1: number of characters in next line.
Line 2: string of switch states.

Sample Output
3

########################################################################
############################ SOLUTION LOGIC ############################
Count the number of adjacent state changes.
One approach is to use regular expression to count number of matches to groups of 0 and 1.
One adjustment in this approach is to remove any trailling 0s because they don't need to be flipped.

Example: 001100
    Correct: 2 steps, manual method.
        001100
        110000
        000000

    Incorrect: 3 steps, regex matching with no adjustment.
        001100
        Match 1: 00
        Match 2: 11
        Match 3: 00

    Correct: 2 steps, regex matching with adjustment (remove trailing 0).
        001100 --> 0011
        Match 1: 00
        Match 2: 11

#########################################################################
############################ MODULE COMMENTS ############################
@author Vincent Zhang
@since 13 July 2021
@context Learning Python 3
"""

# re module allows the use of regular expression. first line of input is not useful for this solution approach.
import re; input()
# Count all matches of one or more 0s and 1s based on input stripped of any trailling 0s.
print(len(re.findall("0+|1+", input().rstrip("0"))))